### PR TITLE
Add debug check for localization parent

### LINKED ISSUE #20406: We should assert that localizations are present
Right now if you use a widget that assumes localizations but don't have a MaterialApp, we crash.
We should be asserting that you have one, with a nice message similar to debugCheckHasMaterial.

### PR DESCRIPTION
**Pull request by jonahwilliams**
There are certain widget which only need localizations in combination with a certain parameter or as a default value. In this case, there might be valid code that would break if this check were added.

While I defaulted to always adding the debug check to the beginning of build - we should probably discuss the best way to do this.

Fixes https://github.com/flutter/flutter/issues/20406

### CODE REVIEW

File: packages/flutter/lib/src/material/debug.dart [hunk 1]
```diff
@@ -66,3 +67,63 @@ bool debugCheckHasMaterial(BuildContext context) {
...
+/// Does nothing if asserts are disabled. Always returns true.
+bool debugCheckHasMaterialLocalizations(BuildContext context) {
+  assert(() {
+    if (Localizations.of<MaterialLocalizations>(context, MaterialLocalizations) == null) {
+      final StringBuffer message = new StringBuffer();
+      message.writeln('No MaterialLocalizations found.');
+      message.writeln(
+          '${context.widget.runtimeType} widgets require MaterialLocalizations '
+              'to be provided by a Localizations widget ancestor.'
+      );
```
Comment (Hixie) [inline packages/flutter/lib/src/material/debug.dart]: indentation is inconsistent here. generally we prefer two spaces indent throughout (as in other parts of the file)

File: packages/flutter/lib/src/material/debug.dart [hunk 2]
```diff
@@ -66,3 +67,63 @@ bool debugCheckHasMaterial(BuildContext context) {
...
+      message.writeln(
+          'Localizations are used to generate many different messages, labels,'
+            'and abbreviations which are used by the material library. '
+      );
+      message.writeln(
+          'To introduce a MaterialLocalizations, either use a'
+              ' MaterialApp at the root of your application to include them '
```
Comment (Hixie) [inline packages/flutter/lib/src/material/debug.dart]: space at the end of the previous line

**Comment by Hixie** [comment]
LGTM module the indentation issue

I recommend volunteering for the next dev roll when you land this.

**Comment by Hixie** [comment]
@jonahwilliams is this still on your radar?

**Comment by jonahwilliams** [comment]
Yup! - got a bit distracted but was planning on landing this week

**Pull request merged**